### PR TITLE
[BIOMAGE-1181] - Add refer a friend button

### PR DESCRIPTION
**Pull request by aerlaut**
# Background
#### Link to issue 

https://biomage.atlassian.net/browse/BIOMAGE-1181

#### Link to staging deployment URL 

https://ui-agi-ui423.scp-staging.biomage.net/

#### Links to any Pull Requests related to this

N.A.

#### Anything else the reviewers should know about the changes here

Add a "Refer a friend" button which reports to the channel #referrals

# Changes
### Code changes

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [X] Unit tests written
- [X] Tested locally with Inframock (with latest production data downloaded with biomage experiment pull)
- [X] Deployed to staging

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

**Comment by aerlaut** [comment]
Hi @vickymorrison can you have a look at this PR? The button is working already. If you join the #referrals channel in Slack, when you create the referral, the message will be created there.

**Review by vickymorrison — APPROVED** [review]

File: src/components/ReferralButton.jsx [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+// Valid email regex based on RFC2822 - https://regexr.com/2rhq7
+const emailRegex = /[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?/;
+
+const ReferralButton = () => {
+  const [visible, setVisible] = useState(false);
+  const [email, setEmail] = useState('');
+  const [isEmailValid, setIsEmailValid] = useState(false);
+  const [customMessage, setCustomMessage] = useState(initialMessage);
+
+  const HOOK_URL = 'aHR0cHM6Ly9ob29rcy5zbGFjay5jb20vc2VydmljZXMvVDAxNTVEWkZWTTAvQjAyQVk0ODQxQ0cvQ0x3Mms4dTBtMkUzcDBVNUhhbjBqeTBv'; // pragma: allowlist secret
```
Comment (StefanBabukov) [inline src/components/ReferralButton.jsx:21]: Do you know what this HOOK_URL is refering to?

File: src/components/ReferralButton.jsx [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+  const submitFeedback = async () => {
```
Comment (StefanBabukov) [inline src/components/ReferralButton.jsx]: you can rename this to submitReferal

File: src/components/ReferralButton.jsx [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+  const overlay = () => (
+    <div>
+      <Card size='small'>
+        <Space direction='vertical' style={{ width: '100%' }}>
+          <Input
+            addonBefore='To: '
+            label='Email'
+            onChange={(e) => {
+              setIsEmailValid(e.target.value.match(emailRegex));
```
Comment (StefanBabukov) [inline src/components/ReferralButton.jsx]: You can use validateInputs here

File: src/components/ReferralButton.jsx [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+import { Auth } from 'aws-amplify';
+import endUserMessages from '../utils/endUserMessages';
+import pushNotificationMessage from '../utils/pushNotificationMessage';
+
+const { TextArea } = Input;
+
+const initialMessage = 'Hi,\n\nCheck out Cellscope from Biomage. It will make your single-cell analysis easier.';
+
+// Valid email regex based on RFC2822 - https://regexr.com/2rhq7
+const emailRegex = /[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?/;
```
Comment (StefanBabukov) [inline src/components/ReferralButton.jsx]: Why don't you validate the email from `utils/validateInputs.js` like you have added VALIDATE_EMAIL at the end of the changes in this PR?

File: src/components/ReferralButton.jsx (hunk 1 above)
Comment (aerlaut) [reply src/components/ReferralButton.jsx:21]: it's the 64-byte encoded form of ReferralBot's [webhook URL](https://api.slack.com/messaging/webhooks). It's encoded to provide more security

File: src/components/ReferralButton.jsx (hunk 2 above)
Comment (aerlaut) [reply src/components/ReferralButton.jsx:25]: oh yea, nice catch. Changed

File: src/components/ReferralButton.jsx (hunk 3 above)
Comment (aerlaut) [reply src/components/ReferralButton.jsx:114]: implemented

File: src/components/ReferralButton.jsx (hunk 4 above)
Comment (aerlaut) [reply src/components/ReferralButton.jsx:15]: hahah yea, added the email check to `validateInputs`

Comment (aerlaut) [reply src/components/ReferralButton.jsx]: hahah yea, it's added now

File: src/components/ReferralButton.jsx (hunk 3 above)
Comment (aerlaut) [reply src/components/ReferralButton.jsx]: implemented

File: src/components/ReferralButton.jsx (hunk 2 above)
Comment (aerlaut) [reply src/components/ReferralButton.jsx]: nice catch! renamed.

File: src/utils/endUserMessages.js [hunk 5]
```diff
@@ -21,6 +21,9 @@ export default {
   ERROR_SIGN_IN: 'We couldn\'t sign you in. Please try again.',
   FEEDBACK_ERROR: 'We couldn\'t process your feedback. Please try submitting again.',
   FEEDBACK_SUCCESSFUL: 'Your feedback has been successfully sent. Thank you!',
+  REFERRAL_ERROR: 'We couldn\'t process your referral. Please try submitting again',
+  REFERRAL_EMAIL_ERROR: 'The error you submitted was invalid. Pleas check again',
```
Comment (StefanBabukov) [inline src/utils/endUserMessages.js]: Fix the typo

**Review by StefanBabukov — APPROVED** [review]
Looks good

File: src/utils/endUserMessages.js (hunk 5 above)
Comment (aerlaut) [reply src/utils/endUserMessages.js]: Fixed. Will merge in after CI has done running.

**Pull request merged**